File: Java/Greedy-Algorithm/Job_Sequencing.java
//JOB SEQUENCING IN JAVA USING GREEDY APPROACH

import java.util.*;

class Job_Sequencing {

    // Declaring variables
    int n; // Size of the array/ No of jobs
    String a[]; // String array to store job numbers
    int b[]; // Integer array to store the profit associated with each job
    int c[]; // Integer array to store the deadline associated with each job

    // Input function to take input from user
    void input() {
        Scanner sc = new Scanner(System.in);
        //taking user input
        System.out.println("Enter number of jobs: ");
        n = sc.nextInt();
        a = new String[n];
        b = new int[n];
        c = new int[n];
        System.out.println("Enter Job No, Profit and Deadline of the job");
        for (int i = 0; i < n; i++) {
            a[i] = sc.next();
            b[i] = sc.nextInt();
            c[i] = sc.nextInt();
        }
        sc.close();

        System.out.println("Arranged Order of Jobs: ");
        System.out.print("Jobs:     ");
        for (int i = 0; i < n; i++) {
            System.out.print(a[i] + " ");
        }
        System.out.println();

        System.out.print("Profit:    ");
        for (int i = 0; i < n; i++) {
            System.out.print(b[i] + " ");
        }
        System.out.println();

        System.out.print("Deadline:   ");
        for (int i = 0; i < n; i++) {
            System.out.print(c[i] + " ");
        }
        System.out.println();
    }

    // Arranging jobs in descending order if not sorted
    void sort() {
        int temp;
        String temp1;
        for (int i = 0; i < n - 1; i++) {
            for (int j = i + 1; j < n; j++) {
                if (b[i] < b[j]) {
                    temp = b[i];
                    b[i] = b[j];
                    b[j] = temp;

                    temp = c[i];
                    c[i] = c[j];
                    c[j] = temp;

                    temp1 = a[i];
                    a[i] = a[j];
                    a[j] = temp1;
                }
            }
        }

        //printing the arranged jobs in descending order

        System.out.println();
        System.out.println("Sorted Order of Jobs: ");
        System.out.print("Jobs:     ");
        for (int i = 0; i < n; i++) {
            System.out.print(a[i] + " ");
        }
        System.out.println();

        System.out.print("Profit:    ");
        for (int i = 0; i < n; i++) {
            System.out.print(b[i] + " ");
        }
        System.out.println();

        System.out.print("Deadline:  ");
        for (int i = 0; i < n; i++) {
            System.out.print(c[i] + " ");
        }
        System.out.println();
    }

    //Function to schedule the jobs and find maximum profit
    void job_sequencing() {
        int max = c[0];
        for (int i = 0; i < n; i++) {
            if (c[i] > max) {
                max = c[i];  // storing the maximum deadline
            }
        }

        String a1[] = new String[max]; //Store the jobs that can yeild highest profit
        //int b1[] = new int[max];
        int profit = 0;

        //finding a free time slot
        for (int i = 0; i < n; i++) {
            int p = c[i];
            p = p - 1;
            if (a1[p] == null) {
                a1[p] = a[i];
                profit += b[i];  //If a slot is found, mark that slot with the job ID, and add its profit to the answer
            } 
            
             //else go to next job and calculate profit
            else {
                while (p != -1) {
                    if (a1[p] == null) {
                        a1[p] = a[i];
                        profit += b[i];
                        break;
                    }
                    p -= 1;
                }
            }
        }

        for (int i = 0; i < max; i++) {
            System.out.print("    " + a1[i]);
        }
        System.out.println();
        System.out.print("Profit Earned " + profit);  // Printing the maximised profit
    }

    //driver code
    public static void main(String args[]) {
        Job_Sequencing ob = new Job_Sequencing();
        ob.input();
        ob.sort();
        ob.job_sequencing();
    }
}

File: Python/Greedy-Algorithm/job_sequencing.py
#!/usr/bin/env python

"""Job Sequencing

This script solves the Job Sequencing problem using Greedy algorithm.
User is prompted to enter:
    1. number of jobs
    2. deadline of each job
    3. profit from each job
The maximum possible profit that can be made after sequencing the jobs
and the optimal job schedule are returned.

Python: 3.10.8

External packages: None

Author: Siddhant Tiwari (github.com/stiwari-ds)
"""

import sys
from typing import Tuple, List


def read_input() -> Tuple[List[int], List[float]]:
    """Reads from standard input, removes leading/trailing whitespace,
    casts the values to their appropriate datatypes, and performs basic
    validation.

    Returns:
        Tuple[List[int], List[float]]:
            1. list of deadlines of each job
            2. list of profits of each job
    """
    try:
        num_jobs = int(input("Number of jobs: ").strip())
    except ValueError:
        sys.exit("Integer value expected.")
    try:
        deadlines = list(map(int, input("Deadlines: ").strip().split()))
    except ValueError:
        sys.exit("Integer values expected.")
    try:
        profits = list(map(float, input("Profits: ").strip().split()))
    except ValueError:
        sys.exit("Numerical values (int/float) expected.")

    if len(deadlines) != num_jobs:
        sys.exit("Deadlines do not match with number of jobs.")
    if len(profits) != num_jobs:
        sys.exit("Profits do not match with number of jobs.")

    return deadlines, profits


def pick_latest_slot(deadline: int, schedule: List[int]) -> int:
    """Picks the latest available slot based on a job's deadline.

    Args:
        deadline (int): deadline for a specific job
        schedule (List[int]): current state of the schedule

    Returns:
        int: chosen slot
    """
    # iterate in reverse over the slots starting from the deadline
    for idx in range(deadline - 1, -1, -1):
        if schedule[idx] is not None:
            continue  # skip the slot if already allotted to a job
        else:
            return idx
    return None  # if no valid slot can be found the job will be skipped


def greedy_schedule(deadlines: List[int], profits: List[float]) -> Tuple[float, str]:
    """Solves the job sequencing problem using a greedy approach. Initiates an
    empty schedule, then iterates through jobs in decresing order of profits, and
    allots latest available slot to each job based on its deadline.

    Args:
        deadlines (List[int]): list of deadlines of each job
        profits (List[float]): list of profits of each job

    Returns:
        Tuple[float, str]:
            1. Total profit from the scheduled jobs
            2. Final schedule after all jobs are considered for scheduling
    """
    num_jobs = len(deadlines)
    schedule = [None] * num_jobs
    total_profit = 0
    # Get indices of jobs in descending order of profits
    sorted_profits_idx = sorted(range(num_jobs), key=profits.__getitem__, reverse=True)
    for job_idx in sorted_profits_idx:
        slot = pick_latest_slot(deadlines[job_idx], schedule)
        if slot is not None:  # if valid slot is returned, update the schedule
            schedule[slot] = job_idx
            total_profit += profits[job_idx]
    # remove empty slots
    schedule = [f"J{j}" for j in schedule if j is not None]
    return total_profit, " -> ".join(schedule)


def job_sequencing():
    deadlines, profits = read_input()
    total_profit, schedule = greedy_schedule(deadlines, profits)
    print(f"\nFinal schedule: {schedule}")
    print(f"Total profit = {total_profit:.2f}\n")


if __name__ == "__main__":
    job_sequencing()
